File: src/main/java/net/dontdrinkandroot/gitki/service/user/RepositoryUserService.java
package net.dontdrinkandroot.gitki.service.user;

import net.dontdrinkandroot.gitki.model.Role;
import net.dontdrinkandroot.gitki.model.User;
import net.dontdrinkandroot.gitki.repository.UserRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.inject.Inject;
import java.util.Iterator;
import java.util.Optional;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
@Service
public class RepositoryUserService implements UserService
{
    private UserRepository userRepository;

    private PasswordEncoder passwordEncoder;

    protected RepositoryUserService()
    {
        /* RI */
    }

    @Inject
    public RepositoryUserService(UserRepository userRepository, PasswordEncoder passwordEncoder)
    {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<User> find(Long id)
    {
        return this.userRepository.findById(id);
    }

    @Override
    @Transactional(readOnly = true)
    public User loadUserByUsername(String username) throws UsernameNotFoundException
    {
        User user = this.userRepository.findByEmail(username);
        if (null == user) {
            throw new UsernameNotFoundException("User not found");
        }

        return user;
    }

    @Override
    @Transactional(readOnly = true)
    public long findCount()
    {
        return this.userRepository.count();
    }

    @Override
    @Transactional(readOnly = true)
    public Iterator<User> find(long first, long count, String property, boolean ascending)
    {
        PageRequest pageRequest = PageRequest.of(
                Math.toIntExact(first / count),
                Math.toIntExact(count),
                ascending ? Sort.Direction.ASC : Sort.Direction.DESC,
                property
        );
        return this.userRepository.findAll(pageRequest).iterator();
    }

    @Override
    @Transactional(readOnly = true)
    public boolean hasAdminUser()
    {
        Long numAdminUsers = this.userRepository.countByRole(Role.ADMIN);
        return numAdminUsers > 0;
    }

    @Override
    @Transactional
    public User save(User user, String password)
    {
        if (null != password) {
            user.setPassword(this.passwordEncoder.encode(password));
        }
        return this.userRepository.save(user);
    }

    @Override
    @Transactional
    public void removeAll()
    {
        this.userRepository.deleteAll();
    }

    @Override
    @Transactional
    public void remove(User user)
    {
        this.userRepository.delete(user);
    }
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/service/user/RepositoryUserService.kt
package net.dontdrinkandroot.gitki.service.user

import net.dontdrinkandroot.gitki.model.Role
import net.dontdrinkandroot.gitki.model.User
import net.dontdrinkandroot.gitki.repository.UserRepository
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.data.repository.findByIdOrNull
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import javax.inject.Inject

@Service
class RepositoryUserService @Inject constructor(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder
) : UserService {

    @Transactional(readOnly = true)
    override fun find(id: Long): User? = userRepository.findByIdOrNull(id)

    @Transactional(readOnly = true)
    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): User = userRepository.findByEmail(username)

    @Transactional(readOnly = true)
    override fun findCount(): Long = userRepository.count()

    @Transactional(readOnly = true)
    override fun find(first: Long, count: Long, property: String, ascending: Boolean): Iterator<User> {
        val pageRequest = PageRequest.of(
            Math.toIntExact(first / count),
            Math.toIntExact(count),
            if (ascending) Sort.Direction.ASC else Sort.Direction.DESC,
            property
        )
        return userRepository.findAll(pageRequest).iterator()
    }

    @Transactional(readOnly = true)
    override fun hasAdminUser(): Boolean = userRepository.countByRole(Role.ADMIN) > 0

    @Transactional
    override fun save(user: User, password: String?): User {
        if (null != password) user.setPassword(passwordEncoder.encode(password))
        return userRepository.save(user)
    }

    @Transactional
    override fun removeAll() = userRepository.deleteAll()

    @Transactional
    override fun remove(user: User) = userRepository.delete(user)
}